extensions-core/caffeine-cache/src/main/java/io/druid/client/cache/CaffeineCache.java

```diff
@@ -0,0 +1,196 @@
+  {
+    return deserialize(cache.getIfPresent(key));
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    cache.put(key, serialize(value));
+  }
+
+  @Override
+  public Map<NamedKey, byte[]> getBulk(Iterable<NamedKey> keys)
+  {
+    // The assumption here is that every value is accessed at least once. Materializing here ensures deserialize is only
+    // called *once* per value.
+    return ImmutableMap.copyOf(Maps.transformValues(cache.getAllPresent(keys), this::deserialize));
+  }
+
+  // This is completely racy with put. Any values missed should be evicted later anyways. So no worries.
+  @Override
+  public void close(String namespace)
+  {
+    if (config.isEvictOnClose()) {
+      cache.asMap().keySet().stream().filter(key -> key.namespace.equals(namespace)).forEach(cache::invalidate);
+    }
+  }
+
+  @Override
+  public io.druid.client.cache.CacheStats getStats()
+  {
+    final com.github.benmanes.caffeine.cache.stats.CacheStats stats = cache.stats();
+    final long size = cache
+        .policy().eviction()
+        .map(eviction -> eviction.isWeighted() ? eviction.weightedSize() : OptionalLong.empty())
+        .orElse(OptionalLong.empty()).orElse(-1);
+    return new io.druid.client.cache.CacheStats(
+        stats.hitCount(),
+        stats.missCount(),
+        cache.estimatedSize(),
+        size,
+        stats.evictionCount(),
+        0,
+        stats.loadFailureCount()
+    );
+  }
+
+  @Override
+  public boolean isLocal()
+  {
+    return true;
+  }
+
+  @Override
+  public void doMonitor(ServiceEmitter emitter)
+  {
+    final CacheStats oldStats = priorStats.get();
+    final CacheStats newStats = cache.stats();
+    final CacheStats deltaStats;
+    if (oldStats == null) {
+      deltaStats = newStats;
+    } else {
+      deltaStats = newStats.minus(oldStats);
+    }
+    final ServiceMetricEvent.Builder builder = ServiceMetricEvent.builder();
+    emitter.emit(builder.build("query/cache/caffeine/delta/requests", deltaStats.requestCount()));
+    emitter.emit(builder.build("query/cache/caffeine/total/requests", newStats.requestCount()));
+    emitter.emit(builder.build("query/cache/caffeine/delta/loadTime", deltaStats.totalLoadTime()));
+    emitter.emit(builder.build("query/cache/caffeine/total/loadTime", newStats.totalLoadTime()));
+    emitter.emit(builder.build("query/cache/caffeine/delta/evictionBytes", deltaStats.evictionWeight()));
+    emitter.emit(builder.build("query/cache/caffeine/total/evictionBytes", newStats.evictionWeight()));
+    if (!priorStats.compareAndSet(oldStats, newStats)) {
+      // ISE for stack trace
+      log.warn(
+          new IllegalStateException("Multiple monitors"),
+          "Multiple monitors on the same cache causing race conditions and unreliable stats reporting"
+      );
+    }
+  }
+
+  Cache<NamedKey, byte[]> getCache()
+  {
+    return cache;
+  }
+
+  private final LZ4Factory factory = LZ4Factory.fastestInstance();
```

Comment:
these could be static and could be at the top of the class instead of down here. Was a little surprised to find them down here as usually declarations are up top.